src/base/StrBuf.cpp

```diff
@@ -0,0 +1,49 @@
+#include "StrBuf.hpp"
+#include <memory.h>
```

Comment:
please include system headers first and then our own.

Reply:
There is a strong reason why a X.c/cpp should include X.h/hpp _first_: makes sure X.h/hpp includes everything it needs without hidden dependencies (include order dependencies are sometime hard to detect and fix)

Example: if X.hpp needs Y.hpp but doesn't include it and X.hpp is the 1st include in X.cpp then compiler early detects this and you get a chance to fix it; otherwise, including Y.hpp (direct or indirect via other headers) will generate no error for the moment but later on, when you forgot about X details and just want to use X.h/hpp again in another context, you'll get errors and you'll have to search what additional headers you have to include in order to use X.h/cpp

Normally there should be no order dependency, headers could be included in any order without implication. If you include the related header first, the compiler can catch order dependency problems. If you put it last, you simply hide possible problems, deferring them for later... do you really want that?

Reply:
Your suggestion makes sense but it does not match with the existing structure. 
Since I don't want to modify the existing files, we can adopt the style for the new files.

Reply:
OK, what about the other headers?
1. related header
2. local/project headers
3. system headers
or
1. related header
2. system headers
3. local/project headers

The first one looks more "natural" from specific/local to general, but let me know how you prefer

Reply:
I prefer to the second, since it does not look too different from the existing style. 
Thank you.

Reply:
@eseokoh & @bsolomenco 

Bogdan's suggestion of ordering included headers from local to global is a general recommendation found in many guides, mainly to avoid hidden dependencies. I would add one more reason that we faced recently: catching define issues and naming conflicts created by our own headers.

My example is ```#define strlen (int) strlen``` in porting.h. New c++ headers use std::strlen, which combined with the our definition of strlen breaks compiler syntax. In some cases system headers come first and the issue is not discovered. However, there is no way to guarantee all system headers are always included before all local headers. One totally unrelated change that impact the headers include order may surface this issue or others like this.

If we'd always include local headers first and then system headers, these kind of issues will be discovered quickly.

We can progressively change the order starting with new files, refactoring files and with files we touch the include section.

Reply:
OK. Then, I change my mind as the first. :)